### ISSUE TITLE
Docker push results in HTTP 405 error when deployed behind ingress-nginx

### ISSUE THREAD

**Issue by brandan-schmitz** [goharbor/harbor-helm#1492]
I have installed Harbor via Helm onto my Kubernetes v1.27.1 cluster and I have been unable to actually push any docker images to the instance. Every time I attempt it I end up with the following error spit out by docker and nothing shows up in the Harbor UI. I have spent 3 days trying so many things and nothing has worked. I feel it is worth noting that I can get it to work if I disable the ingress and set it to be a ClusterIP and then use Istio instead of ingress-nginx. However for this cluster I need to use ingress-nginx. Any help or pointers in figuring out what is going on would be appreciated as I am blocked on a lot of stuff until I get this resolved.

```shell
error parsing HTTP 405 response body: unexpected end of JSON input: ""
```
<br>

### Commands run:
```shell
bschmitz@Brandans-MBP harbor-system % docker login -u admin harbor.somedomain.com
Password: 
Login Succeeded

bschmitz@Brandans-MBP harbor-system % docker pull ubuntu:latest
latest: Pulling from library/ubuntu
dbf6a9befcde: Pull complete 
Digest: sha256:dfd64a3b4296d8c9b62aa3309984f8620b98d87e47492599ee20739e8eb54fbf
Status: Downloaded newer image for ubuntu:latest
docker.io/library/ubuntu:latest

bschmitz@Brandans-MBP harbor-system % docker tag ubuntu:latest harbor.somedomain.com/library/ubuntu:latest

bschmitz@Brandans-MBP harbor-system % docker push harbor.somedomain.com/library/ubuntu:latest
The push refers to repository [harbor.somedomain.com/library/ubuntu]
b8a36d10656a: Pushed 
error parsing HTTP 405 response body: unexpected end of JSON input: ""
```

<br>

### Cluster Information:
- k8s: v1.27.1
- OS: Ubuntu 22.04 LTS
- CNI: Calico
- LoadBalancer provided by MetalLB (L2 mode)
- 8 nodes (3 control plane/5 worker, k8s API is HA via kube-vip)
- Ingress controller: ingress-nginx

<br>

### Ingress NGINX Installation:
I have installed ingress-nginx via its helm chart (v4.6.0) using the following values:
```yaml
controller:
  config:
    custom-http-errors: "400,401,403,404,405,407,408,409,410,411,412,413,416,418,429,500,501,502,503,504,505"
    use-forwarded-headers: "true"
    enable-real-ip: "true"
    proxy-body-size: '0'
    enable-opentracing: "true"
    jaeger-collector-host: jaeger-system-agent.ingress-nginx.svc.cluster.local
    enable-modsecurity: "true"
    enable-owasp-modsecurity-crs: "true"
  extraArgs:
    default-ssl-certificate: ingress-nginx/tls-default-ingress
  ingressClassResource:
    default: true
  kind: DaemonSet
  metrics:
    enabled: true
    serviceMonitor:
      enabled: true
      namespace: monitoring-system
  minReadySeconds: 10
  service:
    annotations:
      metallb.universe.tf/loadBalancerIPs: 10.0.10.254
    externalTrafficPolicy: Local
  updateStrategy:
    rollingUpdate:
      maxUnavailable: 1
    type: RollingUpdate
defaultBackend:
  enabled: true
  extraEnvs:
  - name: TEMPLATE_NAME
    value: lost-in-space
  - name: SHOW_DETAILS
    value: "true"
  image:
    image: tarampampam/error-pages
    registry: ghcr.io
    tag: 2.23.0
  replicaCount: 1
```

<br>

### Harbor Installation:
I have installed Harbor with via this helm chart (v1.12.0) using the following values. Note that I have replaced my domain with `harbor.somedomain.com` for security reasons.
```yaml
# Install Command:
#  - helm upgrade --install harbor harbor/harbor --create-namespace --namespace harbor-system -f harbor-values.yaml --version 1.12.0

expose:
  type: ingress
  tls:
    enabled: true
    certSource: secret
    secret:
      secretName: tls-harbor-ingress
  ingress:
    hosts:
      core: harbor.somedomain.com
    className: nginx
    annotations:
      cert-manager.io/cluster-issuer: letsencrypt-prod
      ingress.kubernetes.io/ssl-redirect: "true"
      ingress.kubernetes.io/proxy-body-size: "0"
      nginx.ingress.kubernetes.io/ssl-redirect: "true"
      nginx.ingress.kubernetes.io/proxy-body-size: "0"

externalURL: https://harbor.somedomain.com

internalTLS:
  enabled: false

ipFamily:
  ipv6:
    enabled: false

persistence:
  enabled: true
  resourcePolicy: keep
  persistentVolumeClaim:
    registry:
      storageClass: ceph-filesystem
      accessMode: ReadWriteMany
      size: 100Gi
    chartmuseum:
      storageClass: ceph-filesystem
      accessMode: ReadWriteMany
      size: 100Gi
    jobservice:
      jobLog:
        storageClass: ceph-filesystem
        accessMode: ReadWriteMany
        size: 10Gi
    database:
      storageClass: ceph-filesystem
      accessMode: ReadWriteMany
      size: 10Gi
    redis:
      storageClass: ceph-filesystem
      accessMode: ReadWriteMany
      size: 10Gi
    trivy:
      storageClass: ceph-filesystem
      accessMode: ReadWriteMany
      size: 100Gi

logLevel: debug

harborAdminPassword: HIDDEN

secretKey: HIDDEN

trivy:
  enabled: false

notary:
  enabled: false

database:
  type: internal
  internal:
    password: hcR9uJZ46KZUKah3

metrics:
  enabled: true
  serviceMonitor:
    enabled: true
```

<br>

### Docker Daemon Debug Logs:
```log
[2023-05-04T04:57:17.155067536Z][dockerd][I] time="2023-05-04T04:57:17.154904192Z" level=debug msg="Calling POST /v1.41/images/ubuntu:latest/tag?repo=harbor.somedomain.com%2Flibrary%2Fubuntu&tag=latest"
[2023-05-04T04:57:33.582604000Z][dockerd][I] time="2023-05-04T04:57:33.582378025Z" level=debug msg="Calling HEAD /_ping"
[2023-05-04T04:57:33.671383947Z][dockerd][I] time="2023-05-04T04:57:33.671037821Z" level=debug msg="Calling POST /v1.41/images/harbor.somedomain.com/library/ubuntu/push?tag=latest"
[2023-05-04T04:57:33.682674459Z][dockerd][I] time="2023-05-04T04:57:33.682479785Z" level=debug msg="hostDir: /etc/docker/certs.d/harbor.somedomain.com"
[2023-05-04T04:57:33.682675804Z][dockerd][I] time="2023-05-04T04:57:33.682643564Z" level=debug msg="Trying to push harbor.somedomain.com/library/ubuntu to https://harbor.somedomain.com v2"
[2023-05-04T04:57:33.730717558Z][dockerd][I] time="2023-05-04T04:57:33.730426340Z" level=debug msg="Pushing repository: harbor.somedomain.com/library/ubuntu:latest"
[2023-05-04T04:57:33.731302929Z][dockerd][I] time="2023-05-04T04:57:33.731057528Z" level=debug msg="Checking for presence of layer sha256:b8a36d10656ac19ddb96ef3107f76820663717708fc37ce929925c36d1b1d157 (sha256:dbf6a9befcdeecbb8813406afbd62ce81394e3869d84599f19f941aa5c74f3d1) in harbor.somedomain.com/library/ubuntu"
[2023-05-04T04:57:33.841746152Z][dockerd][I] time="2023-05-04T04:57:33.841166219Z" level=debug msg="Failed to check for presence of layer sha256:b8a36d10656ac19ddb96ef3107f76820663717708fc37ce929925c36d1b1d157 (sha256:dbf6a9befcdeecbb8813406afbd62ce81394e3869d84599f19f941aa5c74f3d1) in harbor.somedomain.com/library/ubuntu" error="error parsing HTTP 405 response body: unexpected end of JSON input: \"\""
[2023-05-04T04:57:33.841756109Z][dockerd][I] time="2023-05-04T04:57:33.841297390Z" level=debug msg="Pushing layer: sha256:b8a36d10656ac19ddb96ef3107f76820663717708fc37ce929925c36d1b1d157"
[2023-05-04T04:57:34.848760455Z][dockerd][I] time="2023-05-04T04:57:34.848463011Z" level=debug msg="Assembling tar data for 21f184559048a2e5f680927a12db5c6159079a6723780a146ed76aac5d08a8bc"
[2023-05-04T04:57:40.402401623Z][dockerd][I] time="2023-05-04T04:57:40.401737079Z" level=debug msg="uploaded layer sha256:b8a36d10656ac19ddb96ef3107f76820663717708fc37ce929925c36d1b1d157 (sha256:1bc677758ad7fa4503417ae5be18809c5a8679b5b36fcd1464d5a8e41cb13305), 30430220 bytes"
[2023-05-04T04:57:40.443512170Z][dockerd][I] time="2023-05-04T04:57:40.442994732Z" level=info msg="Attempting next endpoint for push after error: error parsing HTTP 405 response body: unexpected end of JSON input: \"\""
```

<br>

### Harbor Registry log file:
[harbor-registry-6987d54ff-lt7rh.log](https://github.com/goharbor/harbor-helm/files/11393065/harbor-registry-6987d54ff-lt7rh.log)

<br>

### Harbor Core log file:
[harbor-core-746f45dd9-xbtxf.log](https://github.com/goharbor/harbor-helm/files/11393086/harbor-core-746f45dd9-xbtxf.log)

<br>

### Summary:
I am unable to push any docker images to any project in Harbor. I can create projects and utilize the UI with no issue and I can login to the docker registry as well as the helm repo for a project but I cannot push a docker image. I have found that if I disable the ingress through ingress-nginx and utilize Isio Gateway/VirtualService/IngressGateway then it works fine however this is not feasible long term for me. When you push an image and you get that error, the image repository does not appear in the Harbor UI, however the quota usage does increase. This seems to suggest that stuff is indeed being pushed out however when it goes to finalize the image something is failing.

Additionally when I login to my Jaeger UI and lookup the traces for these requests everything seems to be working alright except the one request creating the 405 error.
| Order | Method |     Status Result      | Duration | Request URL                                                  |
| :---: | :----: | :--------------------: | :------: | ------------------------------------------------------------ |
|   1   |  GET   |         200 OK         | 49.19ms  | https://harbor.somedomain.com/service/token?account=admin&scope=repository%3Alibrary%2Fubuntu%3Apush%2Cpull&service=harbor-registry |
|   2   |  HEAD  | 405 Method Not Allowed | 15.89ms  | https://harbor.somedomain.com/v2/library/ubuntu/blobs/sha256:dbf6a9befcdeecbb8813406afbd62ce81394e3869d84599f19f941aa5c74f3d1 |
|   3   |  POST  |      202 Accepted      | 979.87ms | https://harbor.somedomain.com/v2/library/ubuntu/blobs/uploads/ |
|   4   | PATCH  |      202 Accepted      |  4.23s   | https://harbor.somedomain.com/v2/library/ubuntu/blobs/uploads/3e1e038b-c855-4e3f-8f6b-46f5cce79098?_state=XNABfIR_JvXumclAGyYLxa4kJ-iUK0Jpt_NSEli8jPl7Ik5hbWUiOiJsaWJyYXJ5L3VidW50dSIsIlVVSUQiOiIzZTFlMDM4Yi1jODU1LTRlM2YtOGY2Yi00NmY1Y2NlNzkwOTgiLCJPZmZzZXQiOjAsIlN0YXJ0ZWRBdCI6IjIwMjMtMDUtMDRUMDQ6NTc6MzQuMTY4ODM2ODE4WiJ9 |
|   5   |  PUT   |      201 Created       | 988.81ms | https://harbor.somedomain.com/v2/library/ubuntu/blobs/uploads/3e1e038b-c855-4e3f-8f6b-46f5cce79098?_state=AsEAcdlw3T84Y_JjHJLZ414cIz3fBab1hlGPPMBvOTl7Ik5hbWUiOiJsaWJyYXJ5L3VidW50dSIsIlVVSUQiOiIzZTFlMDM4Yi1jODU1LTRlM2YtOGY2Yi00NmY1Y2NlNzkwOTgiLCJPZmZzZXQiOjMwNDMwMjIwLCJTdGFydGVkQXQiOiIyMDIzLTA1LTA0VDA0OjU3OjM0WiJ9&digest=sha256%3A1bc677758ad7fa4503417ae5be18809c5a8679b5b36fcd1464d5a8e41cb13305 |
|   6   |  HEAD  |         200 OK         | 178.26ms | https://harbor.somedomain.com/v2/library/ubuntu/blobs/sha256:1bc677758ad7fa4503417ae5be18809c5a8679b5b36fcd1464d5a8e41cb13305 |
|   7   |  HEAD  | 405 Method Not Allowed | 13.95ms  | https://harbor.somedomain.com/v2/library/ubuntu/blobs/sha256:3b418d7b466ac6275a6bfcb0c86fbe4422ff6ea0af444a294f82d3bf5173ce74 |

**Comment by MinerYang** [comment]
Hi @brandan-schmitz 
It is something related to Nginx configurations
maybe you could try add this config or refer to the harbor offline installer internal Nginx settings as well 
https://github.com/distribution/distribution/issues/2386#issuecomment-330199767
https://github.com/goharbor/harbor/blob/main/make/photon/prepare/templates/nginx/nginx.https.conf.jinja

**Comment by brandan-schmitz** [comment]
Unfortunately I have already tried using a server-snipped to manually set the proxy headers to no avail. The configuration I have applied as of last night to the global ingress-nginx and in my Harbor Helm installation will create the same configuration as seen in the template file.

What is confusing me is that all endpoints are working with the exception of the `https://harbor.somedomain.com/v2/library/ubuntu/blobs/sha256:<LAYER_HASH>` endpoint.

Further, the error message seems to indicate that the API is not fully rendering the JSON that is being sent as the error returned is stating that the JSON has been ended unexpectedly. This suggests that it is hitting the endpoint, however somewhere the JSON input is not correct.

**Comment by brandan-schmitz** [comment]
@MinerYang I have enabled nginx's debug-connections option for these requests so I can fully see the headers and what it is routing the requests too internally. From what I can see it is properly populating the Host, X-Real-IP, X-Forwarded-For, X-Forwarded-Proto headers.

```log
"HEAD /v2/library/ubuntu/blobs/sha256:3b418d7b466ac6275a6bfcb0c86fbe4422ff6ea0af444a294f82d3bf5173ce74 HTTP/1.1
uber-trace-id: 7c047af2a139af27:7a24e520fece2c29:7c047af2a139af27:1
Host: harbor.somedomain.com
X-Request-ID: 802dd8278c4f4968c10252db4a41def5
X-Real-IP: 192.168.1.146
X-Forwarded-For: 192.168.1.146
X-Forwarded-Host: harbor.somedomain.com
X-Forwarded-Port: 443
X-Forwarded-Proto: https
X-Forwarded-Scheme: https
X-Scheme: https
User-Agent: docker/20.10.17 go/go1.17.11 git-commit/a89b842 kernel/5.10.104-linuxkit os/linux arch/amd64 UpstreamClient(Docker-Client/20.10.17 \(darwin\))
Authorization: Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6IlJJVUo6WUFNVTpTWUpBOkNNTkQ6NlJWQzpOUUZXOlZUTUM6UUlFUDpNQU9COkc3VVo6RFlGRjpBR0hSIiwidHlwIjoiSldUIn0.eyJpc3MiOiJoYXJib3ItdG9rZW4taXNzdWVyIiwic3ViIjoiYWRtaW4iLCJhdWQiOiJoYXJib3ItcmVnaXN0cnkiLCJleHAiOjE2ODM0MTIyNTMsIm5iZiI6MTY4MzQxMDQ1MywiaWF0IjoxNjgzNDEwNDUzLCJqdGkiOiJUQ0dmN2hUeDNmRlNSOWdtIiwiYWNjZXNzIjpbeyJ0eXBlIjoicmVwb3NpdG9yeSIsIm5hbWUiOiJsaWJyYXJ5L3VidW50dSIsImFjdGlvbnMiOlsicHVsbCIsInB1c2giLCJkZWxldGUiLCIqIl19XX0.uZKH8TIuGkj01PTlBEiQHtTkBk4jbcI-zfAtKGZcQg-HD7_77rjGZbjJ4u-8Xai9qZ169FkVgUI7yr-xUVBD9a78GsFIURU0p9nUTiyulZqVoLuiDoyL6MTzVNNZ3vBA63xtB7z_-rfabZxv6Rn2H9h-esVCZsfCuJyK3MMOr2HbSfRrHH0Qp23BkP1-BJY8wmkr0gAYOQU9csh2UavAtFO7IKNCD-G2t0mTAArcFZlr5FsSLmo_wp_Z9CJ7NuiH4O-2CKT_bAT38shVtLj9iEs-hOqREQtiQy3ovuVXuZK4HqiIJa8S8uyL_mAmcSTnyHdP-lCEBY2soRUJQgBffQ
```

Additionally when it is selecting the backend to route the requests too it is correctly selecting the harbor-core pod. When these requests were made my harbor-core's pod IP was `10.43.230.39` which matches the following in the logs:
```log
2023/05/06 22:00:53 [notice] 1396#1396: *49132 "(.*)" matches "/v2/library/ubuntu/blobs/sha256:3b418d7b466ac6275a6bfcb0c86fbe4422ff6ea0af444a294f82d3bf5173ce74" while sending to client, client: 192.168.1.146, server: harbor.somedomain.com, request: "HEAD /v2/library/ubuntu/blobs/sha256:3b418d7b466ac6275a6bfcb0c86fbe4422ff6ea0af444a294f82d3bf5173ce74 HTTP/1.1", upstream: "http://10.43.230.39:8080/v2/library/ubuntu/blobs/sha256:3b418d7b466ac6275a6bfcb0c86fbe4422ff6ea0af444a294f82d3bf5173ce74", host: "harbor.somedomain.com"
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http script copy: "/"
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http script regex end
2023/05/06 22:00:53 [notice] 1396#1396: *49132 rewritten data: "/", args: "" while sending to client, client: 192.168.1.146, server: harbor.somedomain.com, request: "HEAD /v2/library/ubuntu/blobs/sha256:3b418d7b466ac6275a6bfcb0c86fbe4422ff6ea0af444a294f82d3bf5173ce74 HTTP/1.1", upstream: "http://10.43.230.39:8080/v2/library/ubuntu/blobs/sha256:3b418d7b466ac6275a6bfcb0c86fbe4422ff6ea0af444a294f82d3bf5173ce74", host: "harbor.somedomain.com"
```

Yet the upstream service (the harbor-core pod) is sending back a 404 error:
```log
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http proxy status 404 "404 Not Found"
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http proxy header: "Content-Type: application/json; charset=utf-8"
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http proxy header: "Set-Cookie: sid=faa1e7ed9e7a71d237c55f5f388dfb0e; Path=/; HttpOnly"
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http proxy header: "X-Request-Id: 802dd8278c4f4968c10252db4a41def5"
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http proxy header: "Date: Sat, 06 May 2023 22:00:53 GMT"
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http proxy header: "Content-Length: 133"
2023/05/06 22:00:53 [debug] 1396#1396: *49132 http proxy header done
2023/05/06 22:00:53 [debug] 1396#1396: *49132 finalize http upstream request: 404
```

Note that the IP address `192.168.1.146` is my laptop's address.

**Comment by brandan-schmitz** [comment]
Okay, after further reading through the log file from ingress-nginx I realized when the harness-core pod is returning the 404 error, the default backend is seeing that error and it is manipulating the return error page to conform with the standard error pages for my domain. When it is doing this it ends up returning a different error page than docker is expecting which is causing it to interpret it as a 405 error since the JSON object it is expecting is not there due to ingress-nginx having manipulated the response.

By commenting out the `custom-http-errors` line in the config section of my ingress-nginx helm values file and restarting the ingress-nginx-controllers I am now able to push to the registry.
```yaml
controller:
  config:
    # custom-http-errors: "400,401,403,404,405,407,408,409,410,411,412,413,416,418,429,500,501,502,503,504,505"
    use-forwarded-headers: "true"
    enable-real-ip: "true"
    proxy-body-size: '0'
    enable-opentracing: "true"
    jaeger-collector-host: jaeger-system-agent.ingress-nginx.svc.cluster.local
    enable-modsecurity: "true"
    enable-owasp-modsecurity-crs: "true"
```

However this is not ideal as that then disables the ability for my ingress-nginx deployment to properly display the proper error pages and their associated information since it utilizes that config option to allow nginx to send the error info to the configured defaultBackend.

I did find a workaround to this after looking at the available annotations that ingress-nginx supplies on its ingress class types. By supplying the `nginx.ingress.kubernetes.io/custom-http-errors: "418"` annotation on the harbor ingress I was able to have it override the cluster level configuration. I used 418 since that is not actually utilized by any real services in this instance and so now the ingress-nginx deployment will still be able to utilize the custom-http-errors on the rest of the cluster but it is ignored in essence for the harbor deployment.


# Solution:
In conclusion, when using Harbor with ingress-nginx and you have the defaultBackend enabled and are supplying a `custom-http-errors` in your cluster wide ingress-nginx configuration, you need to add the following annotation to the harbor ingress annotations section:
```yaml
nginx.ingress.kubernetes.io/custom-http-errors: "418"
```
